### DISCUSSION TITLE
next/font works when built locally, but not when deployed to Vercel

### DISCUSSION THREAD

**Discussion by BaileySimrell** [vercel/next.js#48904] [Help]
### Summary

Hello everyone! I could use some help on this one as I'm genuinely stuck.

I'm seeing an odd issue where next/font works perfectly fine in localhost when using both `npm run dev` and `npm run build / start`

Maybe I am missing something obvious here, but I already tried following a few other discussions and configurations like disabling font optimization in the next.config file as well as running `npx update-browserslist-db@latest` which has fixed what seemed like the same error for me just last week, however nothing seems to work in this case for me.. It's also been super difficult trying to debug this as everything in the css inspector / chrome console is identical in the remote vs localhost versions, which has me very confused.

Below is what I have in our tailwind.config.js file
<img width="532" alt="Screenshot 2023-04-27 at 12 06 27 AM" src="https://user-images.githubusercontent.com/6186817/234756962-35a55387-f594-448d-9189-a4ca0cd5353e.png">

And below this is a screenshot of our next.config.js file 
<img width="812" alt="Screenshot 2023-04-27 at 12 06 58 AM" src="https://user-images.githubusercontent.com/6186817/234757059-5283ab57-46bc-4bb0-a02b-76bd6aa98cd7.png">

Here's a Loom video demo of the issue: [https://www.loom.com/share/4ea0967fcdce447099b60b7b4054d6d2](https://www.loom.com/share/4ea0967fcdce447099b60b7b4054d6d2)

And here are screenshots comparing the localhost vs preview branch version: 
<img width="1707" alt="Screenshot 2023-04-27 at 12 10 17 AM" src="https://user-images.githubusercontent.com/6186817/234757406-6d167d4b-d8a5-4e4e-b632-eee433d21153.png">
<img width="1722" alt="Screenshot 2023-04-27 at 12 09 55 AM" src="https://user-images.githubusercontent.com/6186817/234757416-06eee90f-7c1d-4bce-91fd-8c23e1b9689c.png">

### Additional information

```js
Running on next@13.2.3 in app dir
```

### Example

https://web-simrell-slicespace.vercel.app/

**Comment by shyankashani** [comment]
@BaileySimrell I am having the same issue. Were you able to resolve this?

**Comment by shyankashani** [comment]
@BaileySimrell I personally just fixed it by removing font optimization from my `next.config.js` file. Give that a shot?

**Reply by LucasBonafe** [reply]
The `optimizeFonts` function is `true` by default, when setting it to `false` I got the same problem as when deploying to Vercel.

Apparently Vercel defined `false` as the default value, I tried to fix it in a simple way:

![image](https://github.com/vercel/next.js/assets/31395135/569af2af-eeed-4712-b1ff-6de0a60ce655)

And it didn't work.

**Comment by aaort** [comment]
For anyone who faces this issue while using `MUI` library. I had 'use client' directive added to my theme configuration file which caused components that were server-side rendered (e.g. layout) to have invalid font-family class attached to them. everything worked fine after I removed 'use client'.

**Reply by IFurkanCakmak** [reply]
i have never used anywhere use client i am not using mui by the way no ui libs just custom tailwind but font light works in local but not working in vercel deployement. I didnt make any custom configuration in next config. do you have any idea how to solve that?

**Reply by TopIvanAbramov** [reply]
@aaort did you solve the issues?

**Reply by aaort** [reply]
The described solution worked for me while using MUI. I'm working on another project currently and can't provide you guys with additional helpful information. @IFurkanCakmak @TopIvanAbramov

**Reply by JamesDonnelly** [reply]
For anyone calling `createTheme` directly in their theme config file, removing the `'use client'` directive will result in an error.

> Error: Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server". Or maybe you meant to call this function rather than return it.

Two ways to fix this:

- Separate the configuration from the `createTheme` call by moving the `createTheme` call into its own `'use client'` file:

```ts
// theme.ts
export const theme = { ... }
```

```ts
// createTheme.ts
'use client';
import 'theme' from './theme';
export default createTheme(theme);
```

- Alternatively, move the fonts into their own file and import those directly in your layout:

```ts
// fonts.ts
import localFont from "next/font/local"
export const myFont = localFont(...)
```

```ts
// layout.tsx
import './fonts'
```

**Comment by TopIvanAbramov** [comment]
same problem, locally works just fine, Vercel raise an error
<img width="1128" alt="Screenshot 2024-05-14 at 11 32 42" src="https://github.com/vercel/next.js/assets/42469397/ff3943e5-dda0-4c1b-8ded-dc7b74ea7fd6">

**Reply by osas2211** [reply]
Hey @TopIvanAbramov 
Have you been able to find a solution to this issue?

**Reply by anselumjuju** [reply]
@TopIvanAbramov @osas2211 Any solution?, I need to launch my app to production.

**Reply by KBeDevel** [reply]
@anselumjuju try the @MejanH's [solution](https://github.com/vercel/next.js/discussions/48904#discussioncomment-10237026)

**Reply by anselumjuju** [reply]
I solved this by manually adding lightningcss and sharp as explicit dependencies:

```
npm install lightningcss sharp
``` 

Then I triggered a Vercel build without an existing build cache.
This resolved the missing module error during font optimization (next/font + lightningcss issue).

**Reply by MrLFW** [reply]
thanks @anselumjuju, i had the same problem and did `npm install lightningcss` then redeployed without cache (`vercel deploy --prod --force`). didnt need sharp

**Reply by TopIvanAbramov** [reply]
just move to Vite

**Reply by icyJoseph** [reply]
Does this issue happen on v15/16 using next/font, rather strange. Does anyone have a repo where it happens? One shouldn't need to install deps like that.

**Comment by MejanH** [comment]
for me I got it solved by adding `display: "swap",` using tailwindcss. more in here: https://nextjs.org/docs/app/building-your-application/optimizing/fonts#with-tailwind-css

**Reply by KBeDevel** [reply]
After a forced redeploy (without using build cache), this worked.

**Reply by narajaon** [reply]
this is the answer

**Reply by SwanHub** [reply]
legend

**Comment by max-von-aigner** [comment]
i had the same issue where a custom google font was only loading in the local env, but not in the deployed env. 
turns out that i had just forgotten to add the font to the app/layout.tsx file. once i added the const for that custom font to the layout file and redeployed the app, it worked in the deployed env. 
the font was rendering correctly locally, as it was directly imported into globals.css, using tailwind v4.

**Comment by rcortiz** [comment]
I was able to fix mine by adding weights in my fonts.ts file.
